README.md
# TomStick

[![PlatformIO Registry](https://badges.registry.platformio.org/packages/dhanabhon/library/TomStick.svg)](https://registry.platformio.org/libraries/dhanabhon/TomStick) [![License](https://img.shields.io/badge/license-MIT-brightgreen.svg)](https://opensource.org/licenses/MIT)

A TomStick library is an Arduino library for simply controlling the axes of the joystick.

Take control of your joystick with ease using the TomStick library - the ultimate solution for precise and simple axis control on Arduino.

### Example
```c
#include <TomStick.h>

#define JOYSTICK_X_PIN          A0
#define JOYSTICK_Y_PIN          A1

TomStick joystick(JOYSTICK_X_PIN, JOYSTICK_Y_PIN);
TomStick::Direction currentJoystickState = TomStick::UNKNOWN;

void setup() {
    Serial.begin(115200);
    
    // Initialize the joystick values
    joystick.begin();
    // Calibreate the joystick center position
    joystick.calibrate();
    // Set the dead zone of the joystick (default 50)
    joystick.setDeadZoneThreshold(55);
}

void loop() {
    currentJoystickState = joystick.onMove();

    switch (currentJoystickState)
    {
    case TomStick::CENTER:
      Serial.println("CENTER");
      break;
    case TomStick::UP_LEFT:
      Serial.println("UP-LEFT");
      break;
    case TomStick::UP:
      Serial.println("UP");
      break;
    case TomStick::UP_RIGHT:
      Serial.println("UP-RIGHT");
      break;
    case TomStick::LEFT:
      Serial.println("LEFT");
      break;
    case TomStick::RIGHT:
      Serial.println("RIGHT");
      break;
    case TomStick::DOWN_LEFT:
      Serial.println("DONW-LEFT");
      break;
    case TomStick::DOWN:
      Serial.println("DOWN");
      break;
    case TomStick::DOWN_RIGHT:
      Serial.println("DOWN-RIGHT");
      break;
    default:
      Serial.println("UNKNOWN");
      break;
    }
}
```


CHANGELOG.md
# TomStick Arduino Library Changelog

## 1.0.1 - May 27, 2023

- Fix the wording of the dead zone function

## 1.0.0 - Apr 27, 2023

- First version

## 0.0.1 Initial Release


src/TomStick.h
/*
    TomStick.h - Library for controlling the axes of the joystick.
    Created by Tom Dhanabhon, Jan 14, 2023.
*/

#ifndef _____TomStick__
#define _____TomStick__

#include <Arduino.h>

class TomStick {
    public:
        enum Direction {
            UP = 0,
            DOWN = 1,
            LEFT = 2,
            RIGHT = 3,
            UP_RIGHT = 4,
            UP_LEFT = 5,
            DOWN_RIGHT = 6,
            DOWN_LEFT = 7,
            CENTER = 8,
            UNKNOWN = 9
        };

        struct JoystickInfo
        {
            int AxisX;
            int AxisY;
            int DeadZoneThreshold;
        };
        
        TomStick(byte x_pin, byte y_pin);
        void begin(void);

        int readRawAxisX(void);
        int readRawAxisY(void);
        void calibrate(void);
        Direction onMove(void);
        void setDeadZoneThreshold(int);

        JoystickInfo joystickInfo;
        
    private:
        int xValue = 0; // Variable to store X axis value
        int yValue = 0; // Variable to store Y axis value

        int xCenter = 512; // X axis center value
        int yCenter = 512; // Y axis center value

        int threshold = 50; // Joystick movement threshold (deadzone)

        byte X_PIN;
        byte Y_PIN;
};

#endif

src/TomStick.cpp
#include "TomStick.h"

TomStick::TomStick(byte x_pin, byte y_pin) {
    this->X_PIN = x_pin;
    this->Y_PIN = y_pin;
}

void TomStick::begin(void) {
    pinMode(X_PIN, INPUT);
    pinMode(Y_PIN, INPUT);

    joystickInfo.AxisX = 0;
    joystickInfo.AxisY = 0;
    joystickInfo.DeadZoneThreshold = threshold;
}

int TomStick::readRawAxisX(void) {
    return analogRead(X_PIN);
}

int TomStick::readRawAxisY(void) {
    return analogRead(Y_PIN);
}

void TomStick::calibrate() {
    int totalX = 0, totalY = 0;
    int count = 10;
    unsigned long startTime = millis();
    unsigned long elapsedTime = 0;
    int samples = 0;
    while (samples < count) {
        if (elapsedTime >= 50) {
            totalX += analogRead(X_PIN);
            totalY += analogRead(Y_PIN);
            samples++;
            elapsedTime = 0;
        }
        elapsedTime = millis() - startTime;
    }
    xCenter = totalX / count;
    yCenter = totalY / count;
}

/// @brief The threshold value, which represents the deadzone value of the joystick, should be set between 0 and 100.
/// @param threshold 
void TomStick::setDeadZoneThreshold(int threshold = 50) {
    threshold = abs(threshold);

    if (threshold > 512) {
        threshold = 512;
    }

    this->threshold = threshold;
    joystickInfo.DeadZoneThreshold = this->threshold;
}

TomStick::Direction TomStick::onMove(void) {
    Direction currentDirection = UNKNOWN;

    // Read X and Y axis values
    xValue = readRawAxisX();
    yValue = readRawAxisY();

    // Calibrate X and Y axis values
    xValue = xValue - xCenter;
    yValue = yValue - yCenter;

    joystickInfo.AxisX = xValue;
    joystickInfo.AxisY = yValue;

    // Check if joystick has moved beyond threshold
    if (abs(xValue) > threshold || abs(yValue) > threshold) {
        // Detect joystick direction
        if (xValue > threshold && yValue > threshold) {
            currentDirection = DOWN_RIGHT;
        } else if (xValue > threshold && yValue < -threshold) {
            currentDirection = UP_RIGHT;
        } else if (xValue < -threshold && yValue > threshold) {
            currentDirection = DOWN_LEFT;
        } else if (xValue < -threshold && yValue < -threshold) {
            currentDirection = UP_LEFT;
        } else if (xValue > threshold) {
            currentDirection = RIGHT;
        } else if (xValue < -threshold) {
            currentDirection = LEFT;
        } else if (yValue > threshold) {
            currentDirection = DOWN;
        } else if (yValue < -threshold) {
            currentDirection = UP;
        } else {
            currentDirection = UNKNOWN;
        }
    } else {
        currentDirection = CENTER;
    }

    return currentDirection;
}